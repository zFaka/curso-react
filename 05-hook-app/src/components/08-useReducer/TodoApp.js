import React, {useEffect, useReducer} from 'react';
import './styles.css' ;
import {todoReducer} from './todoReducer';
import TodoList from './TodoList';
import TodoAdd from './TodoAdd';

const init = () => {

  return JSON.parse(localStorage.getItem('todos')) || [];
};

const TodoApp = () => {

  const [todos, dispatch] = useReducer(todoReducer, [], init);



  // Should jsonite the objecto to can storage

  useEffect(() => {

    localStorage.setItem('todos', JSON.stringify(todos))
  }, [todos])



  const handleDelete = (todoId) => {


    const delTodoAction = {

      type:'delete', 
      payload:todoId
    };


    dispatch(delTodoAction)
  };



  const handleToggle = (todoId) => {
    dispatch({

      type:'toggle', 
      payload:todoId
    }) 
  }



  const handleAddTodo = (newTodo) => {

    dispatch({

      type:'add', 
      payload:newTodo
    });
  };



  //<p className='text-center complete'>{++i}  {todo.desc}</p>

  return (
    <>
      <h1>TodoApp</h1> 

      <h4>Tasks: {todos.length} </h4>

      <div className='row'>

        <div className=''>
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        </div>

        <div className=''>

          <TodoAdd
            handleAddTodo={handleAddTodo}
          />
        </div>
      </div>

    </>
  )}

export default TodoApp;
