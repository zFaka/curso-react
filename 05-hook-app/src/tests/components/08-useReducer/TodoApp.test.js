import React from 'react';
import {act} from "@testing-library/react";
import {mount, shallow} from "enzyme";
import TodoApp from "../../../components/08-useReducer/TodoApp";
import {demoTodos} from "../../fixtures/demoTodos";

describe('Test TodoApp', () => {


  const wrapper = shallow(<TodoApp />);

  const spy = jest.spyOn(Storage.prototype, 'setItem');


  test('Should render', () => {
    expect(wrapper)
      .toMatchSnapshot()
  })



  test('Should add ToDo', () => {

    const wrapper = mount(<TodoApp/>);

    act(() => {

      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0])
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1])
    })

    
    expect(wrapper.find('h4').text().trim()).toBe('Tasks: 2')
    expect(localStorage.setItem).toHaveBeenCalledTimes(2)
  })




  test('Should del todo', () => {


    act(() => {

      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0])
      wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id)
    })

    
    expect(wrapper.find('h4').text().trim()).toBe('Tasks: 0')
  })
})
