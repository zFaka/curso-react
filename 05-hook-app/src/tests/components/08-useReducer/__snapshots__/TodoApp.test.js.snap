// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TodoApp Should render 1`] = `
<Fragment>
  <h1>
    TodoApp
  </h1>
  <h4>
    Tasks: 
    0
     
  </h4>
  <div
    className="row"
  >
    <div
      className=""
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className=""
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
